diff -urNp a/sub/vobsub.c b/sub/vobsub.c
--- a/sub/vobsub.c	2021-01-24 02:22:09.000000000 +0800
+++ b/sub/vobsub.c	2026-03-15 11:31:19.376308862 +0800
@@ -1285,22 +1285,22 @@ void vobsub_out_output(void *me, const u
            breaks Windows' Vobsub */
         if (vob->fidx && (!last_pts_set || last_pts != pts)) {
             static unsigned int last_h = 9999, last_m = 9999, last_s = 9999, last_ms = 9999;
-            unsigned int h, m, ms;
-            double s;
-            s = pts;
-            h = s / 3600;
-            s -= h * 3600;
-            m = s / 60;
-            s -= m * 60;
-            ms = (s - (unsigned int) s) * 1000;
-            if (ms >= 1000)     /* prevent overflows or bad float stuff */
-                ms = 0;
-            if (h != last_h || m != last_m || (unsigned int) s != last_s || ms != last_ms) {
+            unsigned int h, m, s, ms;
+
+            ms = pts / 90;
+            h = ms / (3600 * 1000);
+            ms -= h * 3600 * 1000;
+            m = ms / (60 * 1000);
+            ms -= m * 60 * 1000;
+            s = ms / 1000;
+            ms %= 1000;
+
+            if (h != last_h || m != last_m || s != last_s || ms != last_ms) {
                 fprintf(vob->fidx, "timestamp: %02u:%02u:%02u:%03u, filepos: %09lx\n",
-                        h, m, (unsigned int) s, ms, ftell(vob->fsub));
+                        h, m, s, ms, ftell(vob->fsub));
                 last_h = h;
                 last_m = m;
-                last_s = (unsigned int) s;
+                last_s = s;
                 last_ms = ms;
             }
         }
